services/src/main/scala/cromwell/services/womtool/impl/WomtoolServiceInCromwellActor.scala

```diff
@@ -0,0 +1,81 @@
+package cromwell.services.womtool.impl
+
+import akka.actor.{ActorRef, Props}
+import cats.data.Validated.{Invalid, Valid}
+import com.typesafe.config.Config
+import cromwell.core.Dispatcher.ServiceDispatcher
+import cromwell.core.WorkflowSourceFilesCollection
+import cromwell.languages.util.ImportResolver.HttpResolver
+import cromwell.languages.LanguageFactory
+import cromwell.languages.util.{ImportResolver, LanguageFactoryUtil}
+import cromwell.services.womtool.WomtoolServiceActor
+import cromwell.services.womtool.WomtoolServiceMessages._
+import cromwell.util.GracefulShutdownHelper.ShutdownCommand
+import wom.core.WorkflowSource
+import wom.expression.NoIoFunctionSet
+
+class WomtoolServiceInCromwellActor(serviceConfig: Config, globalConfig: Config, serviceRegistryActor: ActorRef) extends WomtoolServiceActor {
+
+  override def receive: Receive = {
+    case DescribeRequest(filesCollection) =>
+      sender ! describeWorkflow(filesCollection)
```

Comment:
Avoid blocking calls in receive blocks. IMO it would be best to async this callout to `describeWorkflow` even if that just means stuffing it into an in-memory queue that gets processed one at a time.

Otherwise if we get more than `mailboxSize` requests within the time it takes to make a few http lookups, we'll lose requests

Reply:
>stuffing it into an in-memory queue that gets processed one at a time
Surely with all this advanced tooling lying around we should easily be able to do better? Is creating one `WomtoolServiceInCromwellActor` per request and making them run in parallel an option?

Reply:
There's a balance in there. Having it per request could easily flood the system. There are multiple ways to get some controlled parallelism, e.g. the more typical akka way would be to have a worker pool of a defined size pulling from the queue (but then you have to deal with the question of "what if the queue gets too big"). We handle similar situations with token counting as well. You could also have a specific thread pool and go with your plan (less good as too easy to go awry).

In general I've found that this is the sort of thing which rapidly winds up in premature optimization land (and frequently run afoul of it). One of the nice things about the actor model is that it makes it pretty easy to change the implementation to be more parallel as need be - e.g. replace the current actor w/ a router which is in front of a worker pool and the interface stays the same.

Which is all a long winded way of saying "I wouldn't worry about it just yet" :)

Another thing to consider here is that streaming models (akka or fs2) might be applicable here.

Reply:
Thanks, this is helpful. I’m happy to put exactly as much or as little sophistication into this as we feel is appropriate, just need some guidance to know how

Reply:
Right - there definitely are a whole host of "better" options! All I really care about for now is removing the blocking call in the `receive` method, further optimizations can be done as we find we need them (which may be very soon if we get a lot of timeouts replying to rawls... or it may be never)